# Sheet 9 - Practical Exercise

##  Bushra Tariq Kiyani

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# 1. Loading the Training Data

In [2]:
df = pd.read_csv("synthetic_dataset.csv")
df

Feature_A Feature_B  Feature_C Feature_D  Feature_E Feature_F Feature_G  \
0           4.7       B_3    11475.4       D_2        0.9       F_2       G_1   
1           4.2       B_3    16300.3       D_1        1.3       F_4      G_11   
2           6.2       B_3    13631.4       D_6        1.0       F_2       G_1   
3           1.8       B_3     7405.4       D_3        0.7       F_1       G_6   
4           4.9       B_2    26356.8       D_1        1.3       F_2       G_2   
...         ...       ...        ...       ...        ...       ...       ...   
4495        5.7       B_6    10133.8       D_1        1.3       F_2       G_2   
4496        4.0       B_3    28036.2       D_2        0.9       F_2       G_8   
4497        4.1       B_3    44538.2       D_4        1.4       F_2       G_2   
4498        3.1       B_3    38236.8       D_3        0.7       F_2       G_7   
4499        6.8       B_3    13871.4       D_1        1.3       F_2       G_4   

     Feature_H Feature_I Feature_J  Feature_K  Feature_L  Feature_M Feature_N  \
0          H_2       I_1       J_1        0.0        0.0        4.0       N_1   
1          H_4       I_3       J_2        0.0        0.0        4.0       N_1   
2          H_3       I_1       J_2        0.0        0.0        4.0       N_1   
3          H_4       I_5       J_2        0.0        0.0        2.0       NaN   
4          H_2       I_1       J_1        0.0        0.0        6.0       N_1   
...        ...       ...       ...        ...        ...        ...       ...   
4495       H_2       I_1       J_1        0.0        0.0        2.0       N_1   
4496       H_2       I_1       J_1        0.0        0.0        3.5       N_1   
4497       H_2       I_1       J_1        0.0      197.7        6.5       N_1   
4498       H_2       I_1       J_1        0.0        0.0        4.0       N_1   
4499       H_2       I_1       J_1        0.0        0.0        5.5       N_1   

     Label  
0      Yes  
1      Yes  
2      Yes  
3      Yes  
4       No  
...    ...  
4495   Yes  
4496   Yes  
4497    No  
4498   Yes  
4499    No  

[4500 rows x 15 columns]

# 2. Data Cleaning and Preprocessing
## 2.1  Dealing with Missing Values
We'll check the null values in our dataset.

In [5]:
# Check NaN values
df.info()
df.isna().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4500 entries, 0 to 4499
Data columns (total 15 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Feature_A  4500 non-null   float64
 1   Feature_B  4243 non-null   object 
 2   Feature_C  4500 non-null   float64
 3   Feature_D  4500 non-null   object 
 4   Feature_E  4500 non-null   float64
 5   Feature_F  4500 non-null   object 
 6   Feature_G  4243 non-null   object 
 7   Feature_H  4500 non-null   object 
 8   Feature_I  4500 non-null   object 
 9   Feature_J  4500 non-null   object 
 10  Feature_K  4500 non-null   float64
 11  Feature_L  4500 non-null   float64
 12  Feature_M  4500 non-null   float64
 13  Feature_N  4420 non-null   object 
 14  Label      4500 non-null   object 
dtypes: float64(6), object(9)
memory usage: 527.5+ KB


Feature_A      0
Feature_B    257
Feature_C      0
Feature_D      0
Feature_E      0
Feature_F      0
Feature_G    257
Feature_H      0
Feature_I      0
Feature_J      0
Feature_K      0
Feature_L      0
Feature_M      0
Feature_N     80
Label          0
dtype: int64

In [6]:
df.isnull().any(axis=1).sum()

332

### 2.1.1 Droping rows containing NaN
We see among 4500 observations we have 332 observations with null values. So first we'll implement classification algorithms by droping these 332 observations/rows.

In [7]:
df_DropNa = df.dropna()
df_DropNa.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4168 entries, 0 to 4499
Data columns (total 15 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Feature_A  4168 non-null   float64
 1   Feature_B  4168 non-null   object 
 2   Feature_C  4168 non-null   float64
 3   Feature_D  4168 non-null   object 
 4   Feature_E  4168 non-null   float64
 5   Feature_F  4168 non-null   object 
 6   Feature_G  4168 non-null   object 
 7   Feature_H  4168 non-null   object 
 8   Feature_I  4168 non-null   object 
 9   Feature_J  4168 non-null   object 
 10  Feature_K  4168 non-null   float64
 11  Feature_L  4168 non-null   float64
 12  Feature_M  4168 non-null   float64
 13  Feature_N  4168 non-null   object 
 14  Label      4168 non-null   object 
dtypes: float64(6), object(9)
memory usage: 521.0+ KB


In [8]:
for col in df_DropNa:
    print(df_DropNa[col].nunique())

66
7
3882
16
16
7
14
6
5
2
74
58
75
39
2


In [9]:
for col in df_DropNa:
    print("name of Column is:", col , sorted(df_DropNa[col].unique()))

name of Column is: Feature_A [1.7, 1.8, 1.9, 2.0, 2.1, 2.2, 2.3, 2.4, 2.5, 2.6, 2.7, 2.8, 2.9, 3.0, 3.1, 3.2, 3.3, 3.4, 3.5, 3.6, 3.7, 3.8, 3.9, 4.0, 4.1, 4.2, 4.3, 4.4, 4.5, 4.6, 4.7, 4.8, 4.9, 5.0, 5.1, 5.2, 5.3, 5.4, 5.5, 5.6, 5.7, 5.8, 5.9, 6.0, 6.1, 6.2, 6.3, 6.4, 6.5, 6.6, 6.7, 6.8, 6.9, 7.0, 7.1, 7.2, 7.3, 7.4, 7.5, 7.6, 7.7, 7.9, 8.0, 8.1, 8.4, 9.0]
name of Column is: Feature_B ['B_1', 'B_2', 'B_3', 'B_4', 'B_5', 'B_6', 'B_7']
name of Column is: Feature_C [1921.4, 1939.5, 1984.7, 2010.9, 2029.6, 2046.9, 2169.8, 2204.2, 2205.5, 2231.3, 2296.6, 2307.4, 2343.8, 2364.6, 2394.0, 2400.8, 2410.6, 2418.5, 2472.8, 2476.3, 2582.5, 2582.6, 2593.2, 2595.5, 2624.8, 2629.0, 2635.8, 2675.6, 2689.2, 2690.4, 2695.0, 2703.7, 2704.9, 2716.6, 2724.2, 2738.2, 2740.8, 2740.9, 2744.4, 2748.4, 2766.9, 2777.6, 2788.6, 2803.5, 2815.1, 2822.5, 2828.1, 2829.1, 2833.4, 2836.6, 2841.9, 2857.2, 2864.8, 2885.6, 2905.4, 2917.4, 2925.4, 2931.2, 2959.9, 2965.8, 2966.2, 2967.5, 2969.6, 2976.2, 2977.8, 2981.0, 298

In [10]:
data_Encoded = pd.get_dummies(df_DropNa, columns=["Feature_J", "Label"], drop_first=True)
data_Encoded = pd.get_dummies(data_Encoded)
data_Encoded

Feature_A  Feature_C  Feature_E  Feature_K  Feature_L  Feature_M  \
0           4.7    11475.4        0.9        0.0        0.0        4.0   
1           4.2    16300.3        1.3        0.0        0.0        4.0   
2           6.2    13631.4        1.0        0.0        0.0        4.0   
4           4.9    26356.8        1.3        0.0        0.0        6.0   
5           7.3     2967.5        0.9        0.0        0.0        1.2   
...         ...        ...        ...        ...        ...        ...   
4495        5.7    10133.8        1.3        0.0        0.0        2.0   
4496        4.0    28036.2        0.9        0.0        0.0        3.5   
4497        4.1    44538.2        1.4        0.0      197.7        6.5   
4498        3.1    38236.8        0.7        0.0        0.0        4.0   
4499        6.8    13871.4        1.3        0.0        0.0        5.5   

      Feature_J_J_2  Label_Yes  Feature_B_B_1  Feature_B_B_2  ...  \
0                 0          1              0              0  ...   
1                 1          1              0              0  ...   
2                 1          1              0              0  ...   
4                 0          0              0              1  ...   
5                 1          1              0              0  ...   
...             ...        ...            ...            ...  ...   
4495              0          1              0              0  ...   
4496              0          1              0              0  ...   
4497              0          0              0              0  ...   
4498              0          1              0              0  ...   
4499              0          0              0              0  ...   

      Feature_N_N_37  Feature_N_N_38  Feature_N_N_39  Feature_N_N_4  \
0                  0               0               0              0   
1                  0               0               0              0   
2                  0               0               0              0   
4                  0               0               0              0   
5                  0               0               0              0   
...              ...             ...             ...            ...   
4495               0               0               0              0   
4496               0               0               0              0   
4497               0               0               0              0   
4498               0               0               0              0   
4499               0               0               0              0   

      Feature_N_N_40  Feature_N_N_5  Feature_N_N_6  Feature_N_N_7  \
0                  0              0              0              0   
1                  0              0              0              0   
2                  0              0              0              0   
4                  0              0              0              0   
5                  0              0              0              0   
...              ...            ...            ...            ...   
4495               0              0              0              0   
4496               0              0              0              0   
4497               0              0              0              0   
4498               0              0              0              0   
4499               0              0              0              0   

      Feature_N_N_8  Feature_N_N_9  
0                 0              0  
1                 0              0  
2                 0              0  
4                 0              0  
5                 0              0  
...             ...            ...  
4495              0              0  
4496              0              0  
4497              0              0  
4498              0              0  
4499              0              0  

[4168 rows x 102 columns]

# 3. Spliting into Training + Validation

In [11]:
import sklearn.model_selection
train,validation = sklearn.model_selection.train_test_split(data_Encoded, test_size=0.1,random_state=0)
train.head(5)

Feature_A  Feature_C  Feature_E  Feature_K  Feature_L  Feature_M  \
3937        3.6     6350.9        1.2        0.0        0.0        4.8   
1552        6.0    19118.8        0.9        0.0        0.0        4.0   
1469        3.2     4932.5        0.4        0.0        0.0        4.0   
1275        2.6    42509.2        1.4      217.4        0.0        4.0   
3518        3.6    18897.2        1.6        0.0        0.0        1.0   

      Feature_J_J_2  Label_Yes  Feature_B_B_1  Feature_B_B_2  ...  \
3937              0          0              0              0  ...   
1552              0          1              0              0  ...   
1469              0          1              0              0  ...   
1275              1          1              0              0  ...   
3518              1          1              0              1  ...   

      Feature_N_N_37  Feature_N_N_38  Feature_N_N_39  Feature_N_N_4  \
3937               0               0               0              0   
1552               0               0               0              0   
1469               0               0               0              0   
1275               0               0               0              0   
3518               0               0               0              0   

      Feature_N_N_40  Feature_N_N_5  Feature_N_N_6  Feature_N_N_7  \
3937               0              0              0              0   
1552               0              0              0              0   
1469               0              0              0              0   
1275               0              0              0              0   
3518               0              0              0              0   

      Feature_N_N_8  Feature_N_N_9  
3937              0              0  
1552              0              0  
1469              0              0  
1275              0              0  
3518              0              0  

[5 rows x 102 columns]

# 4. Learning the decision tree

In [12]:
import sklearn.tree
dt = sklearn.tree.DecisionTreeClassifier(min_samples_leaf=50)
X, y = train.drop("Label_Yes",axis=1), train["Label_Yes"]
dt.fit(X, y)

DecisionTreeClassifier(min_samples_leaf=50)

In [13]:
from sklearn.tree import export_graphviz
export_graphviz(dt, "tree.dot", feature_names=X.columns, class_names=["n","y"], filled=True)

In [15]:
#from subprocess import check_call
#check_call(['dot','-Tpng',"tree.dot",'-o','tree.png'])

In [16]:
import sklearn.metrics
print("Decision Tree Training accuracy: ", end="")
print(sklearn.metrics.accuracy_score(train["Label_Yes"], dt.predict(train. drop("Label_Yes",axis=1))))
print("Decision Tree Validation accuracy:", end="")
print(sklearn.metrics.accuracy_score(validation["Label_Yes"], dt.predict(validation.drop("Label_Yes",axis=1))))

Decision Tree Training accuracy: 0.8427086110370567
Decision Tree Validation accuracy:0.8441247002398081
